ambry-replication/src/main/java/com.github.ambry.replication/ReplicaThread.java

```diff
@@ -519,15 +543,19 @@ private void processReplicaMetadataResponse(Set<StoreKey> missingStoreKeys,
             "Blob id is not in the expected partition Actual partition " + blobId.getPartition()
                 + " Expected partition " + remoteReplicaInfo.getLocalReplicaId().getPartitionId());
       }
-      if (!missingStoreKeys.contains(messageInfo.getStoreKey())) {
+      if (!remoteKeyToLocalKeyMap.containsKey(messageInfo.getStoreKey())) {
+        throw new IllegalStateException("remoteKeyToLocalKeyMap does not contain key " + messageInfo.getStoreKey());
+      }
+      BlobId localKey = (BlobId) remoteKeyToLocalKeyMap.get(messageInfo.getStoreKey());
+      if (!missingStoreKeys.contains(messageInfo.getStoreKey()) && localKey != null) {
```

Comment:
As discussed, a few things here:

1. Let us separate out the case where `localKey` is null. The action then would be to remove the corresponding remote key from the `missingRemoteStoreKeys` (so that it is not fetched from the remote).
2. For the notifications, let us be consistent with `AmbryRequests` and use the local key (and not the remote).